### PR TITLE
Try out making a read cache for poolmanager pool modules

### PR DESCRIPTION
**Pull request by ValarDragon**
Try out making a read cache to avoid extra unmarshals.

For the state machine im pretty confident it will work, it may break the errors for some weird queries to old heights (since it will share this cache, and should really return a miss). But this will only be an edge case for new pools.

We should def test it in prod, but I think this should eliminate this 4.73 out of 1100 seconds of sync time: 
image

More importantly i think this will be helpful as a stepping stone to figuring out a more general, re-usable design pattern for us. We could just cache the proto unmarshals, but we would then still pay the key formatting cost. We could cache both key formatting and proto unmarshal in this case.

More general case, we likely do need to cache the two steps differently and use LRU's. 

## Summary by CodeRabbit

- **New Features**
	- Introduced caching for pool module results to enhance performance.
	- Added gas tracking for key-value store operations to improve resource management.

- **Bug Fixes**
	- Improved error handling in pool creation tests by including pool ID context.

- **Refactor**
	- Transitioned method receivers to pointers for consistency and efficiency in pool management modules.

- **Chores**
	- Implemented synchronization mechanisms in pool management to ensure thread safety.
	- Added reset functionality for cache in testing environments to maintain test accuracy.

### CODE REVIEW

File: osmoutils/store_helper.go [hunk 1]
```diff
@@ -284,3 +284,32 @@ func IncreaseCoinByDenomFromPrefix(ctx sdk.Context, storeKey storetypes.StoreKey
...
+	b := store.Get(key)
+	gasVal = uint64(len(b)) * kvGasConfig.ReadCostPerByte
+	if b == nil {
+		return false, gasFlat, gasKey, gasVal, nil
+	}
+	if err := proto.Unmarshal(b, result); err != nil {
+		return true, gasFlat, gasKey, gasVal, err
+	}
+	return true, gasFlat, gasKey, gasVal, nil
+}
```
Comment (czarcas7ic) [inline osmoutils/store_helper.go:309]: Im confused, why do we unmarshal into result if result isn't a pointer or returned?

**Review by PaddyMc — APPROVED** [review]
